src/admin/mod.rs

```diff
@@ -0,0 +1,291 @@
+    pub fn update_ratelimit(
+        ratelimit: Option<Arc<Ratelimiter>>,
+    ) -> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone {
+        warp::path!("ratelimit" / u64)
+            .and(warp::put())
+            .and(with_ratelimit(ratelimit))
+            .and_then(handlers::update_ratelimit)
+    }
+
+    fn with_ratelimit(
+        ratelimit: Option<Arc<Ratelimiter>>,
+    ) -> impl Filter<Extract = (Option<Arc<Ratelimiter>>,), Error = std::convert::Infallible> + Clone
+    {
+        warp::any().map(move || ratelimit.clone())
+    }
+}
+
+pub enum Metric<'a> {
+    Counter(&'a str, Option<&'a str>, u64),
+    Gauge(&'a str, Option<&'a str>, i64),
+    Percentiles(&'a str, Option<&'a str>, Vec<(&'a str, f64, Option<u64>)>),
+}
+
+impl<'a> Metric<'a> {
+    pub fn name(&self) -> &'a str {
+        match self {
+            Self::Counter(name, _description, _value) => name,
+            Self::Gauge(name, _description, _value) => name,
+            Self::Percentiles(name, _description, _percentiles) => name,
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a metriken::MetricEntry> for Metric<'a> {
+    type Error = ();
+
+    fn try_from(metric: &'a metriken::MetricEntry) -> Result<Self, ()> {
+        let any = match metric.as_any() {
+            Some(any) => any,
+            None => {
+                return Err(());
+            }
+        };
+
+        if let Some(counter) = any.downcast_ref::<Counter>() {
+            Ok(Metric::Counter(
+                (*metric).name(),
+                metric.description(),
+                counter.value(),
+            ))
+        } else if let Some(gauge) = any.downcast_ref::<Gauge>() {
+            Ok(Metric::Gauge(
+                metric.name(),
+                metric.description(),
+                gauge.value(),
+            ))
+        } else if let Some(heatmap) = any.downcast_ref::<Heatmap>() {
+            let percentiles = PERCENTILES
+                .iter()
+                .map(|(label, percentile)| {
+                    let value = heatmap.percentile(*percentile).map(|b| b.high()).ok();
+
+                    (*label, *percentile, value)
+                })
+                .collect();
+
+            Ok(Metric::Percentiles(
+                metric.name(),
+                metric.description(),
+                percentiles,
+            ))
+        } else {
+            Err(())
+        }
+    }
+}
+
+mod handlers {
+    use super::*;
+    use core::convert::Infallible;
+    use warp::http::StatusCode;
+
+    pub async fn prometheus_stats() -> Result<impl warp::Reply, Infallible> {
+        let mut data = Vec::new();
+
+        for metric in metriken::metrics()
+            .iter()
+            .map(Metric::try_from)
+            .filter_map(|m| m.ok())
+        {
+            if metric.name().starts_with("log_") {
+                continue;
+            }
+
+            match metric {
+                Metric::Counter(name, description, value) => {
```

Comment:
Can we throw in a link to the exposition standard?
Would this be the right source?  https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md